src/components/ha-assist-chat.ts

```diff
@@ -560,38 +449,125 @@ export class HaAssistChat extends LitElement {
         }
       );
     } catch {
-      hassMessage.text = this.hass.localize("ui.dialogs.voice_command.error");
-      hassMessage.error = true;
-      this.requestUpdate("_conversation");
+      hassMessageProcesser.setError(
+        this.hass.localize("ui.dialogs.voice_command.error")
+      );
     } finally {
       this._processing = false;
     }
   }
 
+  private _createAddHassMessageProcessor() {
+    let currentDeltaRole = "";
+
+    const progressToNextMessage = () => {
+      if (progress.hassMessage.text === "…") {
+        return;
+      }
+      progress.hassMessage.text = progress.hassMessage.text.substring(
+        0,
+        progress.hassMessage.text.length - 1
+      );
+
+      progress.hassMessage = {
+        who: "hass",
+        text: "…",
+        error: false,
+      };
+      this._addMessage(progress.hassMessage);
+    };
+
+    const isAssistantDelta = (
+      _delta: any
+    ): _delta is Partial<ConversationChatLogAssistantDelta> =>
+      currentDeltaRole === "assistant";
+
+    const isToolResult = (
+      _delta: any
+    ): _delta is ConversationChatLogToolResultDelta =>
+      currentDeltaRole === "tool_result";
```

Comment:
I don't get why these are functions here?

Why not just `if (currentDeltaRole === "assistant"`)

Reply:
because TypeScript :(

Reply:
Whenever a "delta" object comes in with a role. That object and all following will be of that type until a new delta object with a role comes in.

Reply:
It's the `_delta is ConversationChatLogToolResultDelta` typescript magic that does the trick.